src/radical/pilot/agent/launch_method/mpirun.py

```diff
@@ -15,114 +15,235 @@ class MPIRun(LaunchMethod):
 
         if '_dplace' in self.name:
-            self._dplace = ru.which('dplace')
-            if not self._dplace:
+            out, err, ret = ru.sh_callout('which dplace', shell=True,
+                                                          env=self._env)
+            dplace = out.strip()
+            if not dplace:
                 raise RuntimeError("dplace not found!")
 
-        self.mpi_version, self.mpi_flavor = \
-                                       self._get_mpi_info(self.launch_command)
+        mpi_version, mpi_flavor = self._get_mpi_info(command)
+
+        # the returned lm_info contains sufficient information to create new,
+        # functional instances of this LM without new initialization
+        self._info = {
+                         'command'    : command,
+                         'dplace'     : dplace,
+                         'ccmrun'     : ccmrun,
+                         'mpt'        : mpt,
+                         'rsh'        : rsh,
+                         'mpi_version': mpi_version,
+                         'mpi_flavor' : mpi_flavor,
+                         'env_sh'     : env_sh
+                     }
+
+        self._init_from_info()
+
+        self._log.debug('===== lm MPIRUN initialize stop: %s', self._info)
+        return self._info
 
 
     # --------------------------------------------------------------------------
     #
-    def construct_command(self, t, launch_script_hop):
+    def finalize(self):
 
-        slots        = t['slots']
-        uid          = t['uid']
-        td          = t['description']
-        sandbox      = t['task_sandbox_path']
-        task_exec    = td['executable']
-        task_threads = td.get('cpu_threads', 1)
-        task_env     = td.get('environment') or dict()
-        task_args    = td.get('arguments')   or list()
-        task_argstr  = self._create_arg_string(task_args)
+        pass
 
-        if '_dplace' in self.name and task_threads > 1:
-            # dplace pinning would disallow threads to map to other cores
-            raise ValueError('dplace can not place threads [%d]' % task_threads)
 
-        # Construct the executable and arguments
-        if task_argstr: task_command = "%s %s" % (task_exec, task_argstr)
-        else          : task_command = task_exec
+    # --------------------------------------------------------------------------
+    #
+    def can_launch(self, task):
+
+        if task['description']['cpu_process_type'] == 'MPI':
+            return True
+
+        return False
 
-        env_string = ''
-        env_list   = self.EXPORT_ENV_VARIABLES + list(task_env.keys())
-        if env_list:
 
-            if self.mpi_flavor == self.MPI_FLAVOR_HYDRA:
-                env_string = '-envlist "%s"' % ','.join(env_list)
+    # --------------------------------------------------------------------------
+    #
+    def get_launcher_env(self):
+
+        return ['. %s' % self._env_sh]
```

Comment:
naming: with this method name I would expect to have a dict of environment variables, maybe something like `get_launcher_env_cmds`?

p.s. when there is no input parameters we can have it as a property:
```python
    @property
    def env_setup_cmds(self):
         return ['. %s' % self._env_sh]
...
# in Popen:
for cmd in launcher.env_setup_cmds:
    ...
```